### PR TITLE
main: Perform a graceful shutdown

### PR DESCRIPTION
**Pull request by doanac**
Copied logic from:

 https://github.com/distribution/distribution/blob/78b9c98c5c31c30d74f9acb7d96f98552f2cf78f/registry/registry.go#L299

Signed-off-by: Andy Doan <andy@foundries.io>

### CODE REVIEW

File: http_graceful.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	}()
+
+	select {
+	case err := <-serveErr:
+		return err
+	case <-quit:
+		// shutdown the server with a grace period of configured timeout
+		c, cancel := context.WithTimeout(ctx, 1*time.Minute)
+		defer cancel()
+		return server.Shutdown(c)
```
Comment (mike-sul) [inline http_graceful.go:34]: The question is whether the `Shutdown` is really "graceful".

**Comment by mike-sul** [comment]
I think the graceful shutdown is not so important at this phase of the EST server development, so LGTM to me.

**Review by mike-sul — APPROVED** [review]

File: http_graceful.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+	"syscall"
+	"time"
+
+	"github.com/labstack/echo/v4"
+)
+
+func RunGracefully(ctx context.Context, server *http.Server, e *echo.Echo) error {
+	// setup channel to get notified on SIGTERM signal
+	quit := make(chan os.Signal, 1)
+	signal.Notify(quit, syscall.SIGTERM)
```
[the remark this answers is not in the document]
Comment (doanac) [reply http_graceful.go:17]: i think the goal is to shutdown gracefully under a normal situation (where k8s tells us to via sigterm). k8s should try that, give us some time, and then forcefully sigkill us. The other signals - i'm not sure they really matter inside production?

[the remark this answers is not in the document]
Comment (mike-sul) [reply http_graceful.go:17]: > under a normal situation (where k8s tells us to via sigterm)

Is using `k8s` new normal :). my point is that we should not make assumptions about execution environment.

It's not critical at all.

**Pull request merged**